### PR TITLE
Feat/add bitstamp connector

### PR DESCRIPTION
**Pull request by Jbekker**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Add a connector for Bitstamp.

**Tests performed by the developer**:
Tested most of the basic functionality like order book snapshots / diffs, placing and canceling orders, trade fills.

**Tips for QA testing**:

### CODE REVIEW

**Comment by Jbekker** [comment]
There are still 3 failing tests, it should be fixed when #7095 is in but I'm open to suggestions.

In short the issue is that ```_request_order_status``` and ```_all_trade_updates_for_order``` are both doing a call to /order_status/, the only difference is omit_transactions=True in the request body. for the order status call. 
```_all_trade_updates_for_order``` is getting called first but the response for it receives is the one without any transaction because that's the one being configured first in the test.

**Comment by cardosofede** [comment]
hey @Jbekker I just approved the workflow. Regarding to your question, with the PR that we discussed last week this should work, but based on this comments I have 2 possible suggestions:
- evaluate if the aioresponses library let's you return the mock based on params like omit_transactions=True (I think that is possible)
- reimplement the test of the base class if necessary

**Comment by Jbekker** [comment]
> hey @Jbekker I just approved the workflow. Regarding to your question, with the PR that we discussed last week this should work, but based on this comments I have 2 possible suggestions:
> 
> * evaluate if the aioresponses library let's you return the mock based on params like omit_transactions=True (I think that is possible)
> * reimplement the test of the base class if necessary

Hi Fede,

* aioresponses only supports returning the mock bases on url / method, also see [https://github.com/pnuckowski/aioresponses/issues/213](https://github.com/pnuckowski/aioresponses/issues/213).
* So you would rather have me reimplement the tests from the base class rather than merging #7095? I don't think that's a bad change, and it also exposed a couple of tests that succeeded because of the wrong reasons.

**Comment by cardosofede** [comment]
@Jbekker merged #7095 let's continue approach!

**Comment by rapcmia** [comment]
NCP approved, https://snapshot.org/#/hbot-ncp.eth/proposal/0xf8f3bd9700ce3fe93d53f0acf1a2aebd50343a32f9bf583b72df7cceef409729

**Comment by rapcmia** [comment]
PR update:
- Use subaccount API key from dev, all ok
- Run market order script ok 
- Setup v1 pmm strategy ok
- Setup pmm_simple controller config and run for 16rs
    - Orders created and cancelled at set executor refresh time 900s
    - Filled order event:
        - Take profit limit close order created and cancelled after reaching time_limit
        - Market close order created and executor closed successfully

---

- XEMM controller w/ another spot connector `in progress`

**Review by rapcmia — APPROVED** [review]
LGTM
- Run tests on local using source and successfully build docker image 
- Use subaccount API key from dev, all ok
- Setup v1 pmm strategy ok
- Setup pmm_simple controller config and run for 16hrs
    - Orders created and cancelled at set executor refresh time 900s
    - Filled order event:
        - Take profit limit close order created and cancelled after reaching time_limit
        - Market close order created and executor closed successfully
- Setup xemm controller config and run for few hours
	- Unfortunately no opportunity for filled order 
	- Orders are cancelled/refreshed when trade profitability is above/below set value 
	- Maker orders are created successfully 
- Run dashboard and launch bot using PR docker image, ok

File: hummingbot/connector/exchange/bitstamp/bitstamp_constants.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+MAX_ORDER_ID_LEN = None
+HBOT_ORDER_ID_PREFIX = "hbot"
+
+# Order States
+ORDER_STATE = {
+    "Open": OrderState.OPEN,
+    "Finished": OrderState.FILLED,
+    "Expired": OrderState.CANCELED,
+    "Canceled": OrderState.CANCELED,
```
Comment (cardosofede) [inline hummingbot/connector/exchange/bitstamp/bitstamp_constants.py:20]: there is not state for failed?

File: hummingbot/connector/exchange/bitstamp/bitstamp_exchange.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+                    trading_pair=order.trading_pair,
+                    fee=fee,
+                    fill_base_amount=amount,
+                    fill_quote_amount=price * amount,
+                    fill_price=price,
+                    fill_timestamp=float(event_data["microtimestamp"]) * 1e-3,
+                )
+                self._order_tracker.process_trade_update(trade_update)
+            else:
+                # These trades don't incur any fees, but they do offset each other.
```
Comment (cardosofede) [inline hummingbot/connector/exchange/bitstamp/bitstamp_exchange.py:355]: I agree with this solution ;)

File: hummingbot/connector/exchange/bitstamp/bitstamp_exchange.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+                fill_quote_amount=Decimal(trade[order.quote_asset.lower()]),
+                fill_price=Decimal(trade["price"]),
+                fill_timestamp=datetime.fromisoformat(trade["datetime"]).timestamp(),
+            )
+            trade_updates.append(trade_update)
+
+        return trade_updates
+
+    async def _request_order_status(self, tracked_order: InFlightOrder) -> OrderUpdate:
+        updated_order_data = await self._api_post(
```
Comment (cardosofede) [inline hummingbot/connector/exchange/bitstamp/bitstamp_exchange.py:464]: this is what I think that we can improve later on and don't ask for each order, instead use the endpoint and collect the updates from the orders returned and if one is not appearing then do the custom request.
not to implement right now but what do you think?

**Review by cardosofede — APPROVED** [review]
Thanks for your contribution and for following the standards @Jbekker.
Great work!

File: hummingbot/connector/exchange/bitstamp/bitstamp_constants.py (hunk 1 above)
Comment (Jbekker) [reply hummingbot/connector/exchange/bitstamp/bitstamp_constants.py:20]: In the documentation I'm seeing only 'Open, Finished, Expired or Canceled' so if there is one, it's undocumented.

**Comment by rapcmia** [comment]
Merged to development and part of release version 2.0.1

**Comment by Jbekker** [comment]
> Hi @Jbekker Please add documentation for this connector to our docs site - https://github.com/hummingbot/hummingbot-site You can use the other connectors as reference. Thanks!

Done! see https://github.com/hummingbot/hummingbot-site/pull/384

**Pull request merged**